src/freenet/support/MultiValueTable.java

```diff
@@ -1,230 +1,260 @@
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            return (v != null) && v.contains(value);
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return false;
         }
+        return list.contains(value);
     }
 
     /**
-     * Users will have to handle synchronizing.
+     * Returns mapped value collection as {@link Enumeration}
+     *
+     * <b>Note:</b> this method is deprecated
+     * please use other {@link #getAllAsList(Object)} method variant, which provides a {@link List}
+     *
+     * @param key key mapping
      */
+    @Deprecated
     public Enumeration<V> getAll(K key) {
-    	Vector<V> v;
-		synchronized (table) {
-			v = table.get(key);
-		}
-        return (v == null ?
-                new EmptyEnumeration<V>() :
-                v.elements());
+        List<V> elements = getAllAsList(key);
+        if (elements.isEmpty()) {
+            return Collections.emptyEnumeration();
+        } else {
+            return Collections.enumeration(elements);
+        }
+    }
+
+    public List<V> getAllAsList(K key) {
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return Collections.emptyList();
+        }
+        return Collections.unmodifiableList(list);
     }
 
     /**
      * To be used in for(x : y).
      */
     public Iterable<V> iterateAll(K key) {
-		synchronized (table) {
-			return(table.get(key));
-		}
+        return getAllAsList(key);
     }
 
     public int countAll(K key) {
-    	Vector<V> v;
-		synchronized (table) {
-			v = table.get(key);
-		}
-    	if(v != null)
-        	return v.size();
-        else
-        	return 0;
+        return getAllAsList(key).size();
     }
 
+    /**
+     * Returns mapped value collection as raw {@link Object}
+     *
+     * <b>Note:</b> this method is deprecated, this was used in previous {@code synchronized} implementation variant,
+     * please use other {@link #getAllAsList(Object)})} method variant, which provides a typed {@link List} collection.
+     *
+     * @param key key mapping
+     */
+    @Deprecated
     public Object getSync(K key) {
-		synchronized (table) {
-			return table.get(key);
-		}
+        return getAllAsList(key);
     }
 
+    /**
+     * Returns mapped value collection as array
+     *
+     * <b>Note:</b> this method is deprecated,
+     * please use other {@link #getAllAsList(Object)} method variant, which provides a typed {@link List} collection.
+     *
+     * @param key key mapping
+     */
+    @Deprecated
     public Object[] getArray(K key) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            if (v == null)
-                return null;
-            else {
-            	Object[] r = new Object[v.size()];
-                v.copyInto(r);
-                return r;
-            }
-        }
+        return getAllAsList(key).toArray();
```

Comment:
This breaks the (undocumented) return value semantics of `getArray`: previously it would return `null` when no values are present, currently it returns an empty array.

Please consider simulating the old behaviour and documenting it in the Javadoc.

Reply:
Restored the previous behavior and put notes into Javadoc